src/samplers/tcp/bpf.c

```diff
@@ -269,3 +272,31 @@ int trace_rto(struct pt_regs *ctx, struct sock *sk)
     add_value(rto.lookup(&loc), 1);
     return 0;
 }
+
+// Run on incoming package validation
+int trace_validate_incoming(struct pt_regs *ctx, struct sock *sk, struct sk_buff *skb) {
+    if (sk == NULL || skb == NULL)
+        return 0;
+
+    struct tcp_sock *tp = tcp_sk(sk);
+    u32 seq = TCP_SKB_CB(skb)->seq;
+    u32 end_seq = TCP_SKB_CB(skb)->end_seq;
+
+    // Segment sequence before the expected one
+    // which means this was a duplicated segment
+    if ((end_seq - tp->rcv_wup) < 0) {
+        // Increment duplicate counter
+        int loc = 0;
+        add_value(duplicate.lookup(&loc), 1);
+    }
+    
+    // Segment sequence after the expected receive window
+    // which means this segment was received out of order
+    u32 window_end = tp->rcv_nxt + tcp_receive_window(tp);
+    if ((window_end - seq) < 0) {
```

Comment:
Similarly here, I think we can simply check whether if seq > tp->rcv_nxt here to see if the segment arrived out of order. 
window_end < seq only tell that the current segment is totally out of the window, but when there's an overlay, it can't capture.